chore(cli): enable Windows CLI E2E tests on GitHub Actions

byCedric opened the pull request:
# Why

Split off from #33204, stacked on #33401 ([for the `EXPO_E2E_TEMP_DIR` workaround](https://github.com/expo/expo/pull/33387#discussion_r1871455847))

This enables running CLI E2E tests on Windows, allowing it to fail until we've merged all fixes required for Windows.

> [!IMPORTANT]
> We should merge this _after_ all Windows fixes have been merged, to ensure we cleared all current Windows pathing issues surfaced from the Jest E2E tests.
> - https://github.com/expo/expo/pull/33401
> - https://github.com/expo/expo/pull/33408
> - https://github.com/expo/expo/pull/33397
> - https://github.com/expo/expo/pull/33396
> - https://github.com/expo/expo/pull/33395
> - https://github.com/expo/expo/pull/33394
> - https://github.com/expo/expo/pull/33393

# How

- Modified **.github/workflows/cli.yml** to include Windows (and allow it to fail).

# Test Plan

- See GitHub Action

# Checklist

<!--
Please check the appropriate items below if they apply to your diff. This is required for changes to Expo modules.
-->

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

byCedric committed:
chore(cli): enable experimental Windows CLI E2E tests on GitHub Actions

Its configured to allow failures, at least until we have merged all fixes for Windows.

byCedric committed:
chore(cli): split CLI e2e workflow into separate Windows workflow

byCedric committed:
chore(cli): fix typo in linux runner

byCedric commented on .github/workflows/cli.yml:
```diff
@@ -42,52 +42,124 @@ concurrency:
+      - name: 🧪 E2E Test CLI (Windows)
+        working-directory: packages/@expo/cli
+        # NOTE(cedric): Windows is significantly slower and suffers from IO congestion when running more than 2 workers
+        run: yarn test:e2e --max-workers 2
```
I've also embedded this into #33401, where it switches tests to run in serial mode when GHA is running in debug mode. Probably can drop the `--max-workers 2` here.

byCedric replied:
Another requirement for this is that we have to modify the temporary system folder on Windows/GHA to be on the same disk as the repository checkout. The issue there is that Metro's FileMap only "scopes" a single disk on Windows [due to its design](https://github.com/facebook/metro/blob/main/packages/metro-file-map/src/lib/TreeFS.js#L96). Meaning that on Windows, it only scopes to `C:\\...` or `D:\\...`.

GHA uses both `C:` (system temp dir) and `D:` (repository checkout) drives, causing `SHA1 not calculated` errors when running Expo CLI from source. I've tested this in: https://github.com/expo/expo/pull/33204/commits/cbe4ca09ccccc28131b8c2a892813394be8bbc91 - which finally got [Windows to pass all E2E tests](https://github.com/expo/expo/actions/runs/12168114455/job/33938213550)

byCedric replied:
Both done!

byCedric committed:
chore: add Windows `EXPO_E2E_TEMP_DIR` workaround to use the same drive as the repositor

This allows us to run Expo CLI from source, without running into the "SHA1 not calculated"-error, caused by a limitation in Metro FileMap

byCedric committed:
chore: remove hardcoded `max-workers` in Windows workflow

byCedric commented on .github/workflows/cli.yml:
```diff
@@ -42,52 +42,125 @@ concurrency:
+        run: yarn test:e2e --shard ${{ matrix.shard }}/${{ strategy.job-total }}
+
+  e2e-windows:
+    runs-on: windows-2022
```
TODO: add sharding once the E2E tests are passing on Windows, to speed things up without unnecessary spamming with error messages.

byCedric commented on .github/workflows/cli.yml:
```diff
@@ -42,52 +42,125 @@ concurrency:
+        run: yarn test:e2e --shard ${{ matrix.shard }}/${{ strategy.job-total }}
+
+  e2e-windows:
+    runs-on: windows-2022
+    continue-on-error: true
```
TODO
- Add sharding once the E2E tests are passing on Windows, to speed things up without unnecessary spamming with error messages.
- Remove the `continue-on-error: true` once the test are passing on Windows

byCedric replied:
```suggestion
    strategy:
      fail-fast: false
      matrix:
        shard: [1, 2, 3, 4]
```

byCedric commented on .github/workflows/cli.yml:
```diff
@@ -42,52 +42,125 @@ concurrency:
 
 jobs:
-  test:
+  e2e-ubuntu:
```
```suggestion
  jest-ubuntu:
```

Maybe this makes more sense, as playwright is also running E2E tests.

byCedric commented on .github/workflows/cli.yml:
```diff
@@ -42,52 +42,125 @@ concurrency:
+        working-directory: packages/@expo/cli
+        run: yarn test:e2e --shard ${{ matrix.shard }}/${{ strategy.job-total }}
+
+  e2e-windows:
```
```suggestion
  jest-windows:
```

byCedric commented on .github/workflows/cli.yml:
```diff
@@ -42,52 +42,125 @@ concurrency:
+        working-directory: packages/@expo/cli
+        run: yarn test:e2e
+  
   web:
```
```suggestion
  playwright-ubuntu:
```

byCedric reviewed (commented):
Maybe this makes more sense, as playwright is also running E2E tests.

byCedric commented:
Ubuntu E2E failure is due to the test timing out and caused by IO congestion, causing the 2x export to take way longer than expected. We could increase the timeout, or limit the workers on ubuntu (and try to speed up using more shards).

Adding max workers has both pro and cons. The tests will execute slower, but the tests will also execute more accurately - where the full power of the machine is available for each test (instead of shared across all sharded tests).

byCedric commented on .github/workflows/cli.yml:
```diff
@@ -42,52 +42,125 @@ concurrency:
 
+      - name: 🧪 E2E Test CLI
+        working-directory: packages/@expo/cli
+        run: yarn test:e2e
```
```suggestion
        run: yarn test:e2e ${{ matrix.shard }}/${{ strategy.job-total }}
```

byCedric reviewed (commented):
TODO
- Add sharding once the E2E tests are passing on Windows, to speed things up without unnecessary spamming with error messages.
- Remove the `continue-on-error: true` once the test are passing on Windows

byCedric reviewed (commented):
-

byCedric committed:
chore: restructure the job names to match the actual executed test suite

byCedric committed:
chore: enable Windows CLI E2E sharding and remove allow failure

brentvatne reviewed (approved):
thanks for putting so much work into this @byCedric! looking forward to #33401 landing so we can start getting these tests running

byCedric committed:
chore: test move all package managers to the D:\ drive

byCedric committed:
test(cli): switch test to serial mode on CI for Windows

Note, going to test the performance on linux before merging. Linux can handle multiple workers, but on Windows it causes Bun to behave inconsistent.

byCedric committed:
test(cli): move `max-workers` config to workflow

Its fast enough with 4 different shards on both Windows and Linux. We can optimize the slow tests a bit more, by reusing the caches. And we can clean up the test output to be more verbose with useful information about the commands executed, but lets do that in a follow-up PR.

byCedric committed:
test(cli): move `max-workers` config to workflow

Its fast enough with 4 different shards on both Windows and Linux. We can optimize the slow tests a bit more, by reusing the caches. And we can clean up the test output to be more verbose with useful information about the commands executed, but lets do that in a follow-up PR.
